include/envoy/config/config_provider.h

```diff
@@ -0,0 +1,98 @@
+#pragma once
+
+#include <memory>
+
+#include "envoy/common/time.h"
+
+#include "common/protobuf/protobuf.h"
+
+#include "absl/types/optional.h"
+
+namespace Envoy {
+namespace Config {
+
+/**
+ * A provider for configuration obtained ether statically or dynamically via xDS APIs.
```

Comment:
One high level question to kick this off with. Is the plan to also replace the static/dynamic treatment of SDS resoureces (e.g. see https://github.com/envoyproxy/envoy/blob/master/source/common/ssl/context_config_impl.cc#L57) with this class framework? It would be helpful to have in mind the concrete types that this will be instantiated with and places used when reading.

Reply:
Yes, SDS looks like a good candidate as well, although I was not planning to transition it to the new API as part of this PR chain so there are likely a few gaps in the generalized API that will need to be addressed.

One of those gaps is that the `ConfigProvider` currently only exposes a `std::shared_ptr<const Config> config() const` function, but the SDS config impls require allowing the user/consumer of the API to set a callback (see [ContextConfigImpl::setSecretUpdateCallback](https://github.com/envoyproxy/envoy/blob/master/source/common/ssl/context_config_impl.h#L53)). From an API standpoint, I think the solution is trivial by exposing a `std::shared_ptr<Config> config()` overload, and having the config impl enforce thread safety itself by leveraging any synchronization mechanisms required.

Reply:
I think it would be ideal from a tech debt perspective if we could at least file an issue for this if not do the conversion at the end of the PR chain. There is a lot of commonality here between RDS and SDS, having multiple methods of managing static/dynamic resources seems redundant. Anyway, I'll do a deep dive now.

Reply:
Agreed, I'll file an issue to keep track of SDS. I plan to take a deeper look at it once I get through these PRs and will likely take on the work myself.

Reply:
Issue filed: #5329.